reader_concurrency_semaphore.cc

```diff
@@ -304,19 +305,34 @@ class reader_permit::impl
     }
 
     db::timeout_clock::time_point timeout() const noexcept {
-        return _timeout;
+        return _timeout_timer.armed() ? _timeout_timer.get_timeout() : db::no_timeout;
     }
 
     void set_timeout(db::timeout_clock::time_point timeout) noexcept {
         using namespace std::chrono_literals;
-        if (_timeout != db::no_timeout && timeout < _timeout) {
-            if (_timeout - timeout > 100ms) {
+        const auto prev_timeout = this->timeout();
+        if (prev_timeout != db::no_timeout && timeout < prev_timeout) {
+            if (prev_timeout - timeout > 100ms) {
                 rcslog.warn("Detected timeout skew of {}ms, please check time skew between nodes in the cluster.  backtrace: {}",
-                        std::chrono::duration_cast<std::chrono::milliseconds>(_timeout - timeout).count(),
+                        std::chrono::duration_cast<std::chrono::milliseconds>(prev_timeout - timeout).count(),
                         current_backtrace());
             }
         }
-        _timeout = timeout;
+        _timeout_timer.cancel();
+        if (timeout != db::no_timeout) {
+            _timeout_timer.set_callback([this] {
+                _abort_source.request_abort_ex(std::make_exception_ptr(timed_out_error{}));
+            });
```

Comment:
This should be done in the timer initializer, since it's the same callback every time it is called.

Reply:
(and it doesn't matter if no_timeout, since it won't be called)